### PR TITLE
New tools

### PR DESCRIPTION
**Pull request by AndersonTorres**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/text/xml/xmldiff/default.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    repo = pname;
+    rev = version;
+    hash = "sha256-xqudHYfwOce2C0pcFzId0JDIIC6R5bllmVKsH+CvTdE=";
+  };
+
+  buildInputs = [
+    lxml
+    six
+  ];
+
```
Comment (SuperSandro2000) [inline pkgs/tools/text/xml/xmldiff/default.nix:24]:
```suggestion

```

File: pkgs/development/tools/castxml/default.nix [hunk 2]
```diff
@@ -1,38 +1,50 @@
-{ lib, stdenv, fetchFromGitHub
-, python3Packages
+{ lib
+, stdenv
+, fetchFromGitHub
+, clang-unwrapped
```
Comment (SuperSandro2000) [inline pkgs/development/tools/castxml/default.nix:4]:
```suggestion
, clang-unwrapped
```

File: pkgs/development/tools/castxml/default.nix [hunk 3]
```diff
@@ -1,38 +1,50 @@
...
-, python3Packages
+{ lib
+, stdenv
+, fetchFromGitHub
+, clang-unwrapped 
 , cmake
-, llvmPackages
-, libffi, libxml2, zlib
-, withMan ? true
+, libclang
```
Comment (SuperSandro2000) [inline pkgs/development/tools/castxml/default.nix:6]:
```suggestion
, libclang
```

**Review by SuperSandro2000 — DISMISSED** [review]

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 116283` at 584cec47 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built successfully:</summary>
  <ul>
    <li><a href="https://gist.github.com/91d1d404d1ccc0d3214f60bdf8185707">castxml</a></li></li>
    <li><a href="https://gist.github.com/89a4e794197a492b668444b6be2b6479">m2r</a></li></li>
    <li><a href="https://gist.github.com/804db506d9dfd53c6dba54f1074bcc66">md4c</a></li></li>
    <li><a href="https://gist.github.com/767f3f2ff1b237b0ce0b8dc4a5c0e144">python38Packages.pygccxml</a></li></li>
    <li><a href="https://gist.github.com/63dafd3ebd8330cf050e96f4d1eef6e5">python39Packages.pygccxml</a></li></li>
    <li><a href="https://gist.github.com/6509c8d9db80c9aebed8311a4d33d33a">xmldiff</a></li></li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

llvm is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/castxml/default.nix:37:3:
```
   |
37 |   buildInputs = [
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/development/tools/castxml/default.nix:52:3:
```
   |
52 |   checkPhase = ''
   |   ^
```
</li>
    <li>warning: unused-argument

Unused argument: `stdenv`.
Near pkgs/tools/text/xml/xmldiff/default.nix:2:3:
```
  |
2 | , stdenv
  |   ^
```
</li>
  </ul>
</details>

File: pkgs/tools/text/m2r/default.nix [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-JNLPEXMoiISh4RnKP+Afj9/PJp9Lrx9UYHsfuGAL7uI=";
+  };
+
+  nativeBuildInputs = [
+    docutils
+  ];
+
+  buildInputs = [
```
Comment (FRidh) [inline pkgs/tools/text/m2r/default.nix:24]: If these need to be available during runtime then they should not be in `buildInputs`.

File: pkgs/tools/text/xml/xmldiff/default.nix [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  version = "2.4";
+
+  src = fetchFromGitHub {
+    owner = "Shoobx";
+    repo = pname;
+    rev = version;
+    hash = "sha256-xqudHYfwOce2C0pcFzId0JDIIC6R5bllmVKsH+CvTdE=";
+  };
+
+  buildInputs = [
```
Comment (FRidh) [inline pkgs/tools/text/xml/xmldiff/default.nix:20]: same here

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -5823,6 +5826,8 @@ in
 
   mcfly = callPackage ../tools/misc/mcfly { };
 
+  m2r = python3Packages.callPackage ../tools/text/m2r { };
```
Comment (FRidh) [inline pkgs/top-level/all-packages.nix:5829]: Avoid using the `callPackage` from another scope. Instead, pass in `python3` or `python3Packages`.

File: pkgs/top-level/all-packages.nix [hunk 7]
```diff
@@ -9291,6 +9296,8 @@ in
 
   xml2 = callPackage ../tools/text/xml/xml2 { };
 
+  xmldiff = python3Packages.callPackage ../tools/text/xml/xmldiff { };
```
Comment (FRidh) [inline pkgs/top-level/all-packages.nix:9299]: same here

**Review by FRidh — CHANGES_REQUESTED** [review]

File: pkgs/top-level/all-packages.nix (hunk 6 above)
Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix:5829]: Oh man.

This is a tool written in Python, using Python tools as dependencies and even using Python traditional tools for packaging. This is not a typical "gcc-configure-make" app.

I can't see why not to use `python3Packages.callPackage` here. Is there some deprecation I am not aware?

File: pkgs/tools/text/m2r/default.nix (hunk 4 above)
Comment (AndersonTorres) [reply pkgs/tools/text/m2r/default.nix:24]: Isn't it the reverse? Things needed only while building are in `nativeBuildInputs` while those needed in run are in`buildInputs`.

File: pkgs/top-level/all-packages.nix (hunk 6 above)
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:5829]: IIRC if you add python to inputs it is a different python which is very hard to debug.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 116283` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
6 packages built:
</summary>
<ul>
<li>castxml</li>
<li>m2r</li>
<li>md4c</li>
<li>python38Packages.pygccxml</li>
<li>python39Packages.pygccxml</li>
<li>xmldiff</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
castxml:
<details>


warning: build-tools-in-build-inputs
llvm is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/development/tools/castxml/default.nix:37:3:

```
   |
37 |   buildInputs = [
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md
</details>
</details>

File: pkgs/top-level/all-packages.nix (hunk 6 above)
Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix:5829]: This is also standard in QT packages. I still remember when I updated openshot.

**Pull request merged**